src/gui/guiEditBox.cpp

```diff
@@ -93,3 +101,727 @@ void GUIEditBox::setAutoScroll(bool enable)
+
+		if (event.KeyInput.Shift) {
+			mark_begin = mb;
+			mark_end = m_cursor_pos;
+		} else {
+			mark_begin = 0;
+			mark_end = 0;
+		}
+
+		return true;
+	}
+
+	// clang-format on
+	return false;
+}
+
+bool GUIEditBox::onKeyDown(const SEvent &event, s32 &mark_begin, s32 &mark_end)
+{
+	// clang-format off
+	if (m_multiline || (m_word_wrap && m_broken_text.size() > 1)) {
+		s32 lineNo = getLineFromPos(m_cursor_pos);
+		s32 mb = (m_mark_begin == m_mark_end) ? m_cursor_pos :
+			(m_mark_begin < m_mark_end ? m_mark_begin : m_mark_end);
+		if (lineNo < (s32)m_broken_text.size() - 1) {
+			s32 cp = m_cursor_pos - m_broken_text_positions[lineNo];
+			if ((s32)m_broken_text[lineNo + 1].size() < cp) {
+				m_cursor_pos = m_broken_text_positions[lineNo + 1] +
+					core::max_((u32)1, m_broken_text[lineNo + 1].size()) - 1;
+			}
+			else
+				m_cursor_pos = m_broken_text_positions[lineNo + 1] + cp;
+		}
+
+		if (event.KeyInput.Shift) {
+			mark_begin = mb;
+			mark_end = m_cursor_pos;
+		} else {
+			mark_begin = 0;
+			mark_end = 0;
+		}
+
+		return true;
+	}
+
+	// clang-format on
+	return false;
+}
+
+void GUIEditBox::onKeyControlC(const SEvent &event)
+{
+	// copy to clipboard
+	if (m_passwordbox || !m_operator || m_mark_begin == m_mark_end)
+		return;
+
+	const s32 realmbgn = m_mark_begin < m_mark_end ? m_mark_begin : m_mark_end;
+	const s32 realmend = m_mark_begin < m_mark_end ? m_mark_end : m_mark_begin;
+
+	core::stringc s;
+	s = Text.subString(realmbgn, realmend - realmbgn).c_str();
+	m_operator->copyToClipboard(s.c_str());
+}
+
+bool GUIEditBox::onKeyControlX(const SEvent &event, s32 &mark_begin, s32 &mark_end)
+{
+	// First copy to clipboard
+	onKeyControlC(event);
+
+	if (m_passwordbox || !m_operator || m_mark_begin == m_mark_end)
+		return false;
+
+	const s32 realmbgn = m_mark_begin < m_mark_end ? m_mark_begin : m_mark_end;
+	const s32 realmend = m_mark_begin < m_mark_end ? m_mark_end : m_mark_begin;
+
+	// Now remove from box if enabled
+	if (isEnabled()) {
+		// delete
+		core::stringw s;
+		s = Text.subString(0, realmbgn);
+		s.append(Text.subString(realmend, Text.size() - realmend));
+		Text = s;
+
+		m_cursor_pos = realmbgn;
+		mark_begin = 0;
+		mark_end = 0;
+		return true;
+	}
+
+	return false;
+}
+
+bool GUIEditBox::onKeyControlV(const SEvent &event, s32 &mark_begin, s32 &mark_end)
+{
+	if (!isEnabled())
+		return false;
+
+	// paste from the clipboard
+	if (!m_operator)
+		return false;
+
+	const s32 realmbgn = m_mark_begin < m_mark_end ? m_mark_begin : m_mark_end;
+	const s32 realmend = m_mark_begin < m_mark_end ? m_mark_end : m_mark_begin;
+
+	// add new character
+	if (const c8 *p = m_operator->getTextFromClipboard()) {
+		if (m_mark_begin == m_mark_end) {
+			// insert text
+			core::stringw s = Text.subString(0, m_cursor_pos);
+			s.append(p);
+			s.append(Text.subString(
+					m_cursor_pos, Text.size() - m_cursor_pos));
+
+			if (!m_max || s.size() <= m_max) // thx to
+							 // Fish
+							 // FH for
+							 // fix
+			{
```

Comment:
```suggestion
			if (!m_max || s.size() <= m_max) {
```

Same below as well.

Reply:
ah strange that the formatter didn't fixed it